code/espurna/settings.cpp

```diff
@@ -6,30 +6,41 @@ Copyright (C) 2016-2019 by Xose Pérez <xose dot perez at gmail dot com>
 
 */
 
-#include "settings.h"
+#include "espurna.h"
 
+#include "crash.h"
 #include "terminal.h"
+#include "storage_eeprom.h"
 
+#include <algorithm>
 #include <vector>
 #include <cstdlib>
 
 #include <ArduinoJson.h>
 
-#include "storage_eeprom.h"
-
 BrokerBind(ConfigBroker);
 
-// -----------------------------------------------------------------------------
-// (HACK) Embedis storage format, reverse engineered
 // -----------------------------------------------------------------------------
 
-unsigned long settingsSize() {
-    unsigned pos = SPI_FLASH_SEC_SIZE - 1;
-    while (size_t len = EEPROMr.read(pos)) {
-        if (0xFF == len) break;
-        pos = pos - len - 2;
-    }
-    return SPI_FLASH_SEC_SIZE - pos + EEPROM_DATA_END;
+namespace settings {
+
+// Depending on features enabled, we may end up with different left boundary
+// Settings are written right-to-left, so we only have issues when there are a lot of key-values
+// XXX: slightly hacky, because we EEPROMr.length() is 0 before we enter setup() code
+kvs_type kv_store(
+    EepromStorage{},
+#if DEBUG_SUPPORT
+    EepromReservedSize + CrashReservedSize,
```

Comment:
@davebuk I also noticed a bug here with the current storage system. Small workaround until I push an update:
```suggestion
   0,
```
Otherwise settings get confused and 'hide' keys after a certain point, so kv might get lost accidentally when value length is the same as the previous one. Easy way to reproduce is to:
```
set one 1
set two 2
set one 2
get two
```
`two` will get lost